gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1268 @@
+
+        # end all processes
+        for p in processes:
+            p.terminate()
+
+        self.asymmetric_distance_matrix = np.concatenate(distances)
+
+        return self.asymmetric_distance_matrix
+
+    def calculate_asymmetric_distance_matrix_chunk(self, ttda1, ttda2, threshold, start_index, method):
+        """Iterates over ttda1 and calculates the
+        distance to every ttda in tttda2.
+
+        Parameters
+        ----------
+        ttda1 and ttda2: 2D arrays of floats
+            Two ttda matrices that are going to be used for distance calculation.
+            Each row in ttda corresponds to one topic. Each cell in the resulting
+            matrix corresponds to the distance between a topic pair.
+        threshold : float, optional
+            threshold defaults to: {"mass": 0.95, "rank": 0.11}, depending on the selected
+            method
+        start_index : number
+            this function might be used in multiprocessing, so start_index has to be set as
+            ttda1 is a chunk of the complete ttda in that case. start_index would be 0
+            if ttda1 == self.ttda. When self.ttda is split into two pieces, each 100 ttdas
+            long, then start_index should be be 100. default is 0
+        method : {'mass', 'rank}, optional
+            method can be "mass" for the original masking method or "rank" for a faster
+            masking method that selects by rank of largest elements in the topic word
+            distribution, to determine which tokens are relevant for the topic.
+
+        Returns
+        -------
+        2D Numpy.numpy.ndarray of floats
+        """
+
+        if method not in ["mass", "rank"]:
+            raise ValueError("method {} unknown".format(method))
+
+        # select masking method:
+        def mass_masking(a):
+            """original masking method. returns a new binary mask"""
+            sorted_a = np.sort(a)[::-1]
+            largest_mass = sorted_a.cumsum() < threshold
+            smallest_valid = sorted_a[largest_mass][-1]
+            return a >= smallest_valid
+
+        def rank_masking(a):
+            """faster masking method. returns a new binary mask"""
+            return a > np.sort(a)[::-1][int(len(a) * threshold)]
+
+        create_mask = {"mass": mass_masking, "rank": rank_masking}[method]
+
+        # some help to find a better threshold by useful log messages
+        avg_mask_size = 0
+
+        # initialize the distance matrix. ndarray is faster than zeros
+        distances = np.ndarray((len(ttda1), len(ttda2)))
+
+        # now iterate over each topic
+        for i in range(len(ttda1)):
+
+            # create mask from a, that removes noise from a and keeps the largest terms
+            a = ttda1[i]
+            mask = create_mask(a)
+            a_masked = a[mask]
+
+            avg_mask_size += mask.sum()
+
+            # now look at every possible pair for topic a:
+            for j in range(len(ttda2)):
```

Comment:
Please look at the itertools module as an alternative to using nested for-loops.

Reply:
For loops changed to enumerator loops for better code comprehension and overall cleaner code.